Sources/SecureXPC/XPCServiceServer.swift

```diff
@@ -0,0 +1,29 @@
+//
+//  File.swift
+//  
+//
+//  Created by Alexander Momchilov on 2021-11-07.
+//
+
+import XPC
+
+public class XPCServiceServer: XPCBaseServer {
+	public static let service = XPCServiceServer()
+
+	private override init() {}
+
+	public override func start() -> Never {
+		xpc_main { connection in
```

Comment:
What happens when this is called not from an XPC Service? Considering it's quite confusing which type of XPC to use, I expect people will screw this up. Would be good if we could throw an error when this is used wrong.

Reply:
Good question. I couldn't find any way to check if a program is executing as an XPC service. Do you know of any way to do that?

Reply:
Searching around a bit I couldn't find a publicly documented way to determine this; however, I think what we want is actually the opposite and that is partially possible enough to be helpful.

It _should be_ possible to determine in the running code is *not* an XPC Service. Namely we can check if the `CFBundlePackageType` Info.plist entry is present and equal to "XPC!". If that's not the case, it's not an XPC Service. (We could go one step further and find the bundle location and see if it's where it should be within the containing app bundle, but not sure that's worth the extra complexity.) When it's not an XPC Service, throwing an error with a useful debug message would be quite helpful.

Reply:
Great find!

Reply:
Yeah I have to imagine the XP Service doesn't work, but I haven't personally tried it. The [documentation](https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingXPCServices.html#//apple_ref/doc/uid/10000172i-SW6-SW6) says this key/value pair is required:
> XPC requires you to specify a number of special key-value pairs in the Info.plist file within the service helper’s bundle. These keys are listed below.

Reply:
I'm opting to use [`CFBundleGetPackageInfo`](https://developer.apple.com/documentation/corefoundation/1537089-cfbundlegetpackageinfo#) for this, because it's public API (rather than relying on the info.plist directly. I doubt the format will change, but why bother risking it).

Here's what that might look like:

```swift
extension Bundle {
	func getPackageInfo() -> (packageType: String?, pacakgeCreator: String?) {
		var packageTypeI = UInt32()
		var packageCreatorI = UInt32()

		CFBundleGetPackageInfo(CFBundleGetMainBundle(), &packageTypeI, &packageCreatorI)

		packageTypeI = packageTypeI.bigEndian
		packageCreatorI = packageCreatorI.bigEndian

		func uint32ToString(_ input: UInt32) -> String? {
			if input == 0 { return nil }
			var input = input
			return String(data: Data(bytes: &input, count: 4), encoding: .utf8)
		}

		return (uint32ToString(packageTypeI), uint32ToString(packageCreatorI))
	}

	func isXPCService() -> Bool {
		self.getPackageInfo().packageType == "XPC!"
	}
}
```

In case you were curious, the actual [implementation of it](https://github.com/apple/swift-corelibs-foundation/blob/512e4121415677c9c85b69eac5a2109815243861/CoreFoundation/PlugIn.subproj/CFBundle_InfoPlist.c#L964-L971) is [totally bizarre](https://github.com/apple/swift-corelibs-foundation/blob/512e4121415677c9c85b69eac5a2109815243861/CoreFoundation/PlugIn.subproj/CFBundle_InfoPlist.c#L850-L958).

Reply:
Using a direct API to access the information is sensible; also reduce the amount of "magic" constants used. This looks great!

I'm not that surprised the implementation for this is so gnarly, that's what a couple decades worth of backwards compatibility and workarounds often end up looking like 😬

If this really only makes sense to be used by the `XPCServiceServer`, seems preferable to not have it be an extension of `Bundle`. This is a reflection of my own architectural bias which is to always default to limited access/scope and only increase it once the need arises; but I don't insist that's the case so long as this remains framework internal.

When this actually make it into the code, would be good to have a comment referencing where the "XPC!" string comes from; namely: https://developer.apple.com/library/archive/documentation/MacOSX/Conceptual/BPSystemStartup/Chapters/CreatingXPCServices.html#//apple_ref/doc/uid/10000172i-SW6-SW6

Nit: `pacakgeCreator` -> `packageCreator`

Reply:
> seems preferable to not have it be an extension of `Bundle`

Ah, I thought it seemed quite fitting. It's what I would do in my own project, but IDK if the rules change once you're writing a library that others integrate, because:

> so long as this remains framework internal.

I've been putting off testing/researching this: Is there UB if two modules define the same internal extensions on the same type? I'll look into it.